File: src/back-end/FunChess/src/FunChess.Core.Chess/Board.cs
using FunChess.Core.Chess.Structs;
using FunChess.Core.Chess.Constants;
using FunChess.Core.Chess.Enums;

namespace FunChess.Core.Chess;

public sealed class Board
{
    public Board()
    {
        InternalBoard = BoardConstants.GenerateDefaultBoard();
        Teams = new Dictionary<Team, DetailedTeam>
        {
            { Team.White, new DetailedTeam(Team.White) },
            { Team.Black, new DetailedTeam(Team.Black) }
        };
    }

    private Board(Cell[] board, Dictionary<Team, DetailedTeam> teams, Team turn)
    {
        InternalBoard = board;
        Teams = teams;
        Turn = turn;
    }

    private MatchState? _lastMatchState;

    internal readonly Cell[] InternalBoard;

    internal readonly Dictionary<Team, DetailedTeam> Teams;

    public Team Turn { get; private set; } = Team.White;

    public Team NextTurn => (Team)((byte)Turn % (byte)Team.Black + (byte)Team.White);
    
    public ReadOnlySpan<Cell> InternalBoardAsSpan() => InternalBoard;

    public bool MovePiece(Move move)
    {
        if (!PieceCanMove(move)) return false;

        InternalBoard[move.Previous.Index].Piece!.Move(this, move);
        ChangeTurn();

        return true;
    }

    public bool PieceCanMove(Move move)
    {
        Cell selectedCell = InternalBoard[move.Previous.Index];
        if (!selectedCell.IsFromTeam(Turn) || move.Previous.Index == move.Next.Index) return false;

        Cell targetCell = InternalBoard[move.Next.Index];
        if (targetCell.IsFromTeam(Turn)) return false;
        
        if (!selectedCell.Piece.MoveIsValid(this, move, out _)) return false;

        InternalBoard[move.Next.Index] = selectedCell;
        InternalBoard[move.Previous.Index] = Cell.Empty;

        bool isKing = selectedCell.Piece == Piece.King;

        if (isKing) Teams[Turn].KingPosition = move.Next;
        bool kingInCheck = KingInCheck();
        if (isKing) Teams[Turn].KingPosition = move.Previous;
        
        InternalBoard[move.Next.Index] = targetCell;
        InternalBoard[move.Previous.Index] = selectedCell;
        
        return !kingInCheck;
    }

    public MatchState GetMatchState()
    {
        if (_lastMatchState.HasValue) return _lastMatchState.Value;
        
        bool kingInCheck = KingInCheck();

        if (!kingInCheck)
        {
            _lastMatchState = HasAtLeastOneMove() ? MatchState.Running : MatchState.Stalemate;
            return _lastMatchState.Value;
        }
        if (HasAtLeastOneMove())
        {
            _lastMatchState = MatchState.Running;
            return _lastMatchState.Value;
        }
        _lastMatchState = (MatchState)NextTurn;
        return _lastMatchState.Value;
    }

    private bool HasAtLeastOneMove()
    {
        for (byte i = 0; i < BoardConstants.TotalSize; i++)
        {
            Cell currentCell = InternalBoard[i];
            if (!currentCell.IsFromTeam(Turn)) continue;

            Position previous = new(i);
            for (byte j = 0; j < BoardConstants.TotalSize; j++)
            {
                Move possibleMove = new(previous, new Position(j));
                if (PieceCanMove(possibleMove)) return true;
            }
        }
        return false;
    }

    public bool KingInCheck()
    {
        Position kingPosition = GetKingPosition();
        Position? currentTeamExposedEnPassant = Teams[Turn].ExposedEnPassant;
        Position? targetTeamExposedEnPassant = Teams[NextTurn].ExposedEnPassant;
        CastlingPlay targetCastlingPlays = Teams[NextTurn].CastlingPlays;
        for (byte i = 0; i < BoardConstants.TotalSize; i++)
        {
            Cell currentCell = InternalBoard[i];
            if (currentCell.IsEmpty() || currentCell.IsFromTeam(Turn)) continue;

            Move possibleMove = new(new Position(i), kingPosition);

            ChangeTurn();
            bool moveIsValid = currentCell.Piece.MoveIsValid(this, possibleMove, out _);

            Teams[Turn].ExposedEnPassant = targetTeamExposedEnPassant;
            Teams[Turn].CastlingPlays = targetCastlingPlays;
            Turn = NextTurn;
            Teams[Turn].ExposedEnPassant = currentTeamExposedEnPassant;

            if (moveIsValid) return true;
        }
        return false;
    }

    private Position GetKingPosition()
    {
        Position? kingPosition = Teams[Turn].KingPosition;
        if (kingPosition.HasValue) return kingPosition.Value;
        
        Cell kingCell = Cell.Get(Piece.King, Turn);
        for (byte i = 0; i < BoardConstants.TotalSize; i++)
        {
            if (InternalBoard[i] == kingCell) return new Position(i);
        }
        throw new Exception("Invalid board! Missing king.");
    }

    private void ChangeTurn()
    {
        _lastMatchState = null;
        Turn = NextTurn;
        Teams[Turn].ExposedEnPassant = null;
    }

    public byte[] ToByteArray()
    {
        List<byte> bytes = new() { (byte)Turn };
        for (int i = 1; i <= Teams.Count; i++)
        {
            bytes.Add((byte)Teams[(Team)i].CastlingPlays);
            bytes.Add(Teams[(Team)i].ExposedEnPassant?.Index ?? byte.MaxValue);
        }
        
        Cell lastCell = InternalBoard[0], currentCell = InternalBoard[0];
        int repeatCount = 1;
        
        for (int i = 1; i < BoardConstants.TotalSize; i++)
        {
            currentCell = InternalBoard[i];
            if (lastCell == currentCell) { ++repeatCount; continue; }

            AddCellsInfoToBytes(bytes, lastCell, repeatCount);
            repeatCount = 1;
            lastCell = currentCell;
        }
        AddCellsInfoToBytes(bytes, currentCell, repeatCount);
        return bytes.ToArray();
    }
    
    private static void AddCellsInfoToBytes(ICollection<byte> bytes, Cell cell, int repeatCount)
    {
        bytes.Add((byte)repeatCount);
        if (cell.IsEmpty())
        {
            bytes.Add(byte.MinValue);
            return;
        }
        bytes.Add((byte)cell);
    }

    public override string ToString()
    {
        ReadOnlySpan<byte> bytes = ToByteArray();
        return Convert.ToBase64String(bytes);
    }

    public static Board Parse(byte[] bytes)
    {
        var turn = (Team)bytes[0];
        var teams = new Dictionary<Team, DetailedTeam>(2);

        int bytesIndex = 1;
        for (int i = 1; i <= 2; i++)
        {
            teams[(Team)i] = new DetailedTeam
            (
                (CastlingPlay)bytes[bytesIndex], 
                bytes[bytesIndex + 1] == byte.MaxValue ? null : new Position(bytes[bytesIndex + 1])
            );
            bytesIndex += 2;
        }

        var board = new Cell[BoardConstants.TotalSize];
        for (int i = 0; bytesIndex < bytes.Length; bytesIndex += 2)
        {
            int repeatCount = bytes[bytesIndex];
            for (int j = 0; j < repeatCount; j++) board[i++] = (Cell)bytes[bytesIndex + 1];
        }
        return new Board(board, teams, turn);
    }
}

File: src/front-end/fun-chess-app/src/core/chess/Board.ts
import Cell from "./Cell";
import DetailedTeam from "./DetailedTeam";
import { BOARD_TOTAL_SIZE, getInitialBoard } from "./constants/board-constants";
import CastlingPlay from "./enums/CastlingPlay";
import MovementResult from "./enums/MovementResult";
import SpecialMove from "./enums/SpecialMove";
import Team from "./enums/Team";
import King from "./pieces/King";
import Move from "./structs/Move";
import Position from "./structs/Position";

export default class Board {
    public constructor(board?: Cell[], teams?: Map<Team, DetailedTeam>, team?: Team) {
        if (board === undefined || teams === undefined || team === undefined) {
            this.internalBoard = getInitialBoard();
            this.teams = new Map<Team, DetailedTeam>();
            this.teams.set(Team.White, new DetailedTeam(undefined, undefined, Team.White));
            this.teams.set(Team.Black, new DetailedTeam(undefined, undefined, Team.Black));
            
            return;
        }
        this.internalBoard = board;
        this.teams = teams;
        this._turn = team
    }

    private _turn = Team.White;

    public get turn(): Team {
        return this._turn;
    }

    public get nextTurn(): Team {
        return this._turn % Team.Black + Team.White;
    }

    public readonly internalBoard: Cell[];

    public readonly teams: Map<Team, DetailedTeam>;

    public movePiece(move: Move): MovementResult {
        const movementResult = this.pieceCanMove(move);
        if (movementResult === MovementResult.None || movementResult === MovementResult.Illegal) return movementResult;

        this.internalBoard[move.previous.index].piece!.move(this, move);
        this.changeTurn();
        
        if (this.kingInDanger()) return MovementResult.Check;
        return movementResult;
    }

    public pieceCanMove(move: Move): MovementResult {
        const selectedCell = this.internalBoard[move.previous.index];
        if (selectedCell.isEmpty() || move.previous.index == move.next.index || !selectedCell.isFromTeam(this._turn)) return MovementResult.None;

        const targetCell = this.internalBoard[move.next.index];
        if (!targetCell.isEmpty() && targetCell.isFromTeam(this._turn)) {
            if (this.kingInDanger()) return MovementResult.Illegal;
            return MovementResult.None;
        }
        
        const moveIsValidTuple = selectedCell.piece!.moveIsValid(this, move);
        if (!moveIsValidTuple[0]) {
            if (this.kingInDanger()) return MovementResult.Illegal;
            return MovementResult.None;
        }

        const kingAlreadyInDanger = this.kingInDanger();

        this.internalBoard[move.next.index] = selectedCell;
        this.internalBoard[move.previous.index] = Cell.Empty;

        const isKing = selectedCell.piece === King.instance;
        if (isKing) {
            this.teams.get(this._turn)!.kingPosition = move.next;
        }
        const kingInDanger = this.kingInDanger();
        if (isKing) {
            this.teams.get(this._turn)!.kingPosition = move.previous;
        }
        
        this.internalBoard[move.next.index] = targetCell;
        this.internalBoard[move.previous.index] = selectedCell;
        
        if (kingInDanger) return kingAlreadyInDanger ? MovementResult.Illegal : MovementResult.None;
        if (moveIsValidTuple[1] === SpecialMove.LeftCastling || moveIsValidTuple[1] === SpecialMove.RightCastling) return MovementResult.Castle;
        if (targetCell.isEmpty()) return MovementResult.Move;

        return MovementResult.Capture;
    }

    public kingInDanger(): boolean {
        const kingPosition = this.getKingPosition();
        const currentTeamExposedEnPassant = this.teams.get(this._turn)!.exposedEnPassant;
        const targetTeamExposedEnPassant = this.teams.get(this.nextTurn)!.exposedEnPassant;
        const targetCastlingPlays = this.teams.get(this.nextTurn)!.castlingPlays;
        for (let i = 0; i < BOARD_TOTAL_SIZE; i++) {
            const currentCell = this.internalBoard[i];
            if (currentCell.isEmpty() || currentCell.isFromTeam(this._turn)) continue;

            const possibleMove = new Move(new Position(i), kingPosition);

            this.changeTurn();
            const moveIsValid = currentCell.piece!.moveIsValid(this, possibleMove)[0];

            this.teams.get(this._turn)!.exposedEnPassant = targetTeamExposedEnPassant;
            this.teams.get(this._turn)!.castlingPlays = targetCastlingPlays;
            this._turn = this.nextTurn;
            this.teams.get(this._turn)!.exposedEnPassant = currentTeamExposedEnPassant;

            if (moveIsValid) return true;
        }
        return false;
    }

    private getKingPosition(): Position {
        const kingPosition: Position | null = this.teams.get(this._turn)!.kingPosition;
        if (kingPosition !== null) return kingPosition;

        const kingCell = Cell.get(King.instance, this._turn)
        for (let i = 0; i < BOARD_TOTAL_SIZE; i++) {
            if (this.internalBoard[i] === kingCell) return new Position(i);
        }
        throw new Error("Invalid board! Missing king.");
    }

    private changeTurn(): void {
        this._turn = this.nextTurn;
        this.teams.get(this._turn)!.exposedEnPassant = null;
    }

    public static Parse(bytes: Uint8Array): Board {
        const turn: Team = bytes[0];
        const teams = new Map<Team, DetailedTeam>();

        let bytesIndex = 1;
        for (let i = 1; i <= 2; i++) {
            const exposedEnPassant = bytes[bytesIndex + 1] === 255 ? null : new Position(bytes[bytesIndex + 1]);
            const castlingPlays: CastlingPlay = bytes[bytesIndex];

            teams.set(i, new DetailedTeam(castlingPlays, exposedEnPassant));
            bytesIndex += 2;
        }

        const board: Cell[] = [];
        for (let i = 0; bytesIndex < bytes.length; bytesIndex += 2) {
            let repeatCount = bytes[bytesIndex];
            for (let j = 0; j < repeatCount; j++) board[i++] = Cell.parse(bytes[bytesIndex + 1]);
        }
        return new Board(board, teams, turn);
    }
}